src/eddymotion/model/dipy.py

```diff
@@ -0,0 +1,417 @@
+
+        Parameters
+        ----------
+        gtab : :obj:`~dipy.core.gradients.GradientTable`
+            The gradient table with which the GP has been fitted.
+        model: :obj:`~sklearn.gaussian_process.GaussianProcessRegressor`
+            The fitted Gaussian process regressor object.
+        mask: :obj:`~numpy.ndarray`
+            The boolean mask used during fitting (can be ``None``).
+
+        """
+        self.fitted_gtab = gtab
+        self.model = model
+        self.mask = mask
+
+    def predict(
+        self,
+        gtab: GradientTable,
+    ) -> np.ndarray:
+        """
+        Generate DWI signal based on a fitted Gaussian Process.
+
+        Parameters
+        ----------
+        gtab: :obj:`~dipy.core.gradients.GradientTable`
+            Gradient table with one or more orientations at which the GP will be evaluated.
+
+        Returns
+        -------
+        :obj:`numpy.ndarray`
+            A 3D or 4D array with the simulated gradient(s).
+
+        """
+        return gp_prediction(self.fitted_gtab, gtab, self.model, mask=self.mask)
+
+
+def _ensure_positive_scale(
+    a: float,
+) -> None:
+    if a <= 0:
+        raise ValueError(f"a must be strictly positive. Provided: {a}")
+
+
+def compute_exponential_covariance(
+    theta: np.ndarray,
+    a: float,
+) -> np.ndarray:
+    r"""Compute the exponential covariance matrix following eq. 9 in [Andersson15]_.
+
+    .. math::
+
+        C(\theta) = \exp(- \frac{\theta}{a})
+
+    Parameters
+    ----------
+    theta : :obj:`~numpy.ndarray`
+        Pairwise angles across diffusion gradient encoding directions.
+    a : :obj:`float`
+        Positive scale parameter that here determines the "distance" at which θ
+        the covariance goes to zero.
+
+    Returns
+    -------
+    :obj:`~numpy.ndarray`
+        Exponential covariance function values.
+
+    """
+
+    _ensure_positive_scale(a)
+
+    return np.exp(-theta / a)
+
+
+def compute_spherical_covariance(
+    theta: np.ndarray,
+    a: float,
+) -> np.ndarray:
+    r"""Compute the spherical covariance matrix following eq. 10 in [Andersson15]_.
+
+    .. math::
+
+        C(\theta) = \begin{cases}
+            1 - \frac{3 \theta}{2 a} + \frac{\theta^3}{2 a^3} & \textnormal{if} \; \theta \leq a \\
+            0 & \textnormal{if} \; \theta > a
+        \end{cases}
+
+    Parameters
+    ----------
+    theta : :obj:`~numpy.ndarray`
+        Pairwise angles across diffusion gradient encoding directions.
+    a : :obj:`float`
+        Positive scale parameter that here determines the "distance" at which θ
+        the covariance goes to zero.
+
+    Returns
+    -------
+    :obj:`~numpy.ndarray`
+        Spherical covariance matrix.
+
+    """
+
+    _ensure_positive_scale(a)
+
+    return np.where(theta <= a, 1 - 3 * (theta / a) ** 2 + 2 * (theta / a) ** 3, 0)
+
+
+def compute_pairwise_angles(
+    gtab_X: GradientTable,
+    gtab_Y: GradientTable | None = None,
```

Comment:
Since this will probably make it into dipy one day, my point of having these being ndarrays does not stand. Better use the gradient table object, which gives us two goodies:

- implicit checking of dimensionality (and therefore, dropped checks within the implementation)
- access to the b-values (which will come handy when we support multi-shell kernels).

Reply:
:+1: Resolving.